core/jvm/src/main/scala/cats/effect/unsafe/WorkStealingThreadPool.scala

```diff
@@ -398,7 +400,7 @@ private[effect] final class WorkStealingThreadPool(
 
     if (thread.isInstanceOf[WorkerThread]) {
       val worker = thread.asInstanceOf[WorkerThread]
-      if (worker.isOwnedBy(pool)) {
+      if (worker.isOwnedBy(pool) && !allWorkersParked) {
```

Comment:
```suggestion
      if (worker.isOwnedBy(pool) && !notifyShouldWakeup()) {
```

The idea here is sound. Let's hope the benchmarks agree with us.

Reply:
Thanks, I have been looking into it now. I don't think we want to use the external queue only if `((st & UnparkMask) >>> UnparkShift) < threadCount`. On the contrary, I'm getting the best results when using the condition `((st & UnparkMask) >>> UnparkShift) == treadCount` instead, which also makes the most sense to me (I.e. no threads are searching, all workers are unparked, means computation heavy fibers). Doing so does not only increase fairness, but preliminary results show that it increases throughput in existing benchmarks.

Reply:
For reference, this is the fairness with `!notifyShouldWakeup()`

![image](https://user-images.githubusercontent.com/44302748/140667878-5d5e80fc-84fc-49b6-b034-c54557b48e00.png)

This is the fairness with (now) `!allWorkersBusy()`

![image](https://user-images.githubusercontent.com/44302748/140667897-f510f5fb-3079-4524-8694-156a611a686e.png)

Preliminary results show an increase in performance, in addition to the increase in fairness.